static/js/publisher/form/media.js

```diff
@@ -0,0 +1,225 @@
+import React, { Fragment } from "react";
+import ReactDOM from "react-dom";
+
+import { PropTypes } from "prop-types";
+
+import MediaItem, { mediaClasses } from "./mediaItem";
+
+class Media extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.markForDeletion = this.markForDeletion.bind(this);
+    this.addImage = this.addImage.bind(this);
+    this.mediaChange = this.mediaChange.bind(this);
+    this.keyboardEvent = this.keyboardEvent.bind(this);
+    this.focused = this.focused.bind(this);
+    this.blurred = this.blurred.bind(this);
+
+    this.state = {
+      mediaData: props.mediaData,
+      focused: null
+    };
+  }
+
+  componentDidUpdate(prevProps, prevState) {
+    this.props.updateState(prevState.mediaData);
+    if (
+      prevState.mediaData.filter(item => item.status !== "delete").length >
+      this.props.mediaLimit
+    ) {
+      // Do the query lookup only when we need it - otherwise every state
+      // update will trigger a DOM lookup
+      const warning = document.querySelector(".js-media-limit-warning");
+      if (warning) {
+        warning.classList.remove("u-hide");
+      }
+    } else {
+      const warning = document.querySelector(".js-media-limit-warning");
+      if (warning) {
+        warning.classList.add("u-hide");
+      }
+    }
+
+    if (
+      this.state.focused === "new" &&
+      document.activeElement !== this.blankItem
+    ) {
+      this.blankItem.focus();
+    }
+  }
+
+  markForDeletion(key) {
+    let index = -1;
+    let newFocus = this.state.focused;
+
+    const newMediaData = this.state.mediaData.map((item, i) => {
+      if (item.url === key) {
+        item.status = "delete";
+        index = i;
+      }
+      return item;
+    });
+
+    if (this.state.focused && index > -1) {
+      const available = newMediaData.filter(item => item.status !== "delete");
+      if (available[0]) {
+        newFocus = available[0].url;
+      } else {
+        newFocus = "new";
+      }
+    }
+
+    this.setState({
+      mediaData: newMediaData,
+      focused: newFocus
+    });
+  }
+
+  mediaChange(input) {
+    const newMediaData = this.state.mediaData;
+    for (let i = 0; i < input.files.length; i++) {
+      const file = input.files[i];
+      newMediaData.push({
+        file,
+        url: URL.createObjectURL(file),
+        name: file.name,
+        type: "screenshot",
+        status: "new"
+      });
+    }
+
+    this.setState({
+      mediaData: newMediaData
+    });
+  }
+
+  addImage() {
+    const input = document.createElement("input");
+    input.type = "file";
+    input.accept = "image/x-png,image/gif,image/jpeg";
+    input.name = "screenshots";
+    input.hidden = "hidden";
+    input.addEventListener("change", () => {
+      this.mediaChange(input);
+    });
+
+    const form = ReactDOM.findDOMNode(this).closest("form"); // eslint-disable-line react/no-find-dom-node
```

Comment:
It's better to access DOM of a component using ["refs"](https://reactjs.org/docs/refs-and-the-dom.html).

Reply:
can you use a ref for an element outside of the ReactDOM?

Reply:
No. You can store a ref to elements rendered by React. But at least this would allow replacing `findDOMNode` if I read it correctly.

Reply:
Ah, I see. Fixed!